### PR TITLE
ROX-12948 add port pool in DeploymentExposure tests

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

Changes the tests to retry with a different port if applying the service fails due to the port being already allocated.

Depends on:
* #4729

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* Tested with a local cluster and an infra cluster. Test passed even with all the services having a port collision.

### CODE REVIEW

File: sensor/tests/resource/service/service_test.go [hunk 1]
```diff
@@ -185,71 +187,157 @@ func (s *DeploymentExposureSuite) Test_ClusterIpPermutation() {
 	)
 }
 
+var nextPort int32 = 30007
```
Comment (lvalerom) [inline sensor/tests/resource/service/service_test.go:190]: Set this to 30000.

File: sensor/tests/resource/service/service_test.go [hunk 2]
```diff
@@ -185,71 +187,157 @@ func (s *DeploymentExposureSuite) Test_ClusterIpPermutation() {
...
-							ContainerPort: 80,
-							Exposure:      storage.PortConfig_NODE,
-							ExposureInfos: []*storage.PortConfig_ExposureInfo{
-								{
-									ServiceName: fmt.Sprintf("nginx-svc-node-port-%d", c.port),
-									ServicePort: 80,
-									NodePort:    c.port,
-									Level:       storage.PortConfig_NODE,
+		var err, portErr error
+		port := c.port
```
Comment (lvalerom) [inline sensor/tests/resource/service/service_test.go:290]: Use `getPort` here and generate the service here.

File: sensor/tests/resource/service/service_test.go [hunk 3]
```diff
@@ -450,66 +477,183 @@ func (s *DeploymentExposureSuite) Test_NodePortPermutationWithPod() {
...
+			},
+		},
+	}
+}
+
+func setDynamicFields(svc *v1.Service, name string, port int32, sel map[string]string, serviceFunc func(*v1.Service, string, int32, map[string]string)) {
+	serviceFunc(svc, name, port, sel)
+}
+
+func setDynamicFieldsInSlice(resources []resource.K8sResourceInfo, portConfig []*storage.PortConfig, name string, port int32, sel map[string]string, svcFunc serviceFunc, pcFunc portConfigFunc) {
```
Comment (fredrb) [inline sensor/tests/resource/service/service_test.go:652]: I think naming can be improved here. I suggest using the suffix `fn` for functions passed as parameter. 

Also, the parameter name `pcFunc` is a bit cryptic, could we call it as it is, something like: `portConfigFn`? Same for `svcFunc`.

**Review by fredrb — APPROVED** [review]
Nice that you were able to remove the service yaml files. I think this looks better now. I don't see any blockers if the tests are green. I do think some variable names could be improved though.

File: sensor/tests/resource/service/service_test.go (hunk 3 above)
Comment (lvalerom) [reply sensor/tests/resource/service/service_test.go:652]: Done

**Comment by lvalerom** [comment]
> /test gke-nongroovy-e2e-tests

This failure was unrelated to the changes; re-triggering to make sure the changed tests pass.

**Pull request merged**